solver/src/solver/http_solver.rs

```diff
@@ -171,40 +174,75 @@ impl HttpSolver {
         result
     }
 
-    fn map_constant_product_orders_for_solver(
-        &self,
-        orders: Vec<ConstantProductOrder>,
-    ) -> HashMap<usize, ConstantProductOrder> {
+    fn map_amm_orders_for_solver<T>(&self, orders: Vec<T>) -> HashMap<usize, T> {
         orders.into_iter().enumerate().collect()
     }
 
     fn amm_models(
         &self,
-        amms: &HashMap<usize, ConstantProductOrder>,
+        constant_product_orders: &HashMap<usize, ConstantProductOrder>,
+        weighted_product_orders: &HashMap<usize, WeightedProductOrder>,
         gas_price: f64,
     ) -> HashMap<usize, PoolModel> {
         let uniswap_cost = self.uniswap_cost(gas_price);
-        amms.iter()
+        let mut pool_model_map = HashMap::new();
+        let constant_product_models: HashMap<_, PoolModel> = constant_product_orders
+            .iter()
             .map(|(index, amm)| {
                 let mut reserves = HashMap::new();
                 reserves.insert(amm.tokens.get().0, U256Wrapper(U256::from(amm.reserves.0)));
                 reserves.insert(amm.tokens.get().1, U256Wrapper(U256::from(amm.reserves.1)));
 
-                let uniswap = PoolModel {
+                let pool_model = PoolModel {
                     parameters: PoolParameters::ConstantProduct(ConstantProductPoolParameters {
-                        fee: BigDecimal::from(*amm.fee.numer())
-                            / BigDecimal::from(*amm.fee.denom()),
                         reserves,
                     }),
+                    fee: BigDecimal::from(*amm.fee.numer()) / BigDecimal::from(*amm.fee.denom()),
                     cost: CostModel {
                         amount: uniswap_cost,
                         token: self.native_token,
                     },
                     mandatory: false,
                 };
-                (*index, uniswap)
+                (*index, pool_model)
             })
-            .collect()
+            .collect();
+        // TODO - get actual balancer cost (or at least an upper bound)
+        let balancer_cost = self.uniswap_cost(gas_price);
+        let weighted_product_models: HashMap<_, PoolModel> = weighted_product_orders
+            .iter()
+            .map(|(index, amm)| {
+                let reserves = amm
+                    .reserves
+                    .iter()
+                    .map(|(token, state)| {
+                        (
+                            *token,
+                            PoolTokenData {
+                                balance: state.balance,
+                                weight: BigDecimal::from_usize(usize::from(state.weight))
+                                    .expect("failed to convert Bfp to BigDecimal"),
```

Comment:
I think the function `u256_to_big_int` might help here, I had the same issue [in my PR](https://github.com/gnosis/gp-v2-services/pull/755/files#diff-38988351762381d171eb4c77de97eaed54429cff6fe6fa178b166f50e9fc243fR117-R126).

Reply:
Cool, although this will now become part of another PR.